web/src/main/java/org/springframework/security/web/csrf/GenerateTokenProvider.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.web.csrf;
+
+/**
+ * Functional interface to provide CSRF token generation logic
+ *
+ * @author Ruby Hartono
+ *
+ * @param <T> the type of the returned CsrfToken
+ * @since 5.4
+ */
+@FunctionalInterface
+public interface GenerateTokenProvider<T extends CsrfToken> {
```

Comment:
We shouldn't need this interface.

Reply:
I think you missed this comment. We shouldn't need this interface. We should use a `Producer<CsrfToken>` instead.

Reply:
We still need this interface. As mentioned from the other comments.

Do you mean to rename this from `GenerateTokenProvider<T extends CsrfToken>` to `Producer<CsrfToken>` ?

Reply:
Ahh sorry. I meant to say`Supplier<CsrfToken.

All occurrences of `GenerateTokenProvider` can be replaced with a `Supplier<CsrfToken>`.

Reply:
do you mean `java.util.function.Supplier` ?

nope, it cannot be replaced. `GenerateTokenProvider` accepts 3 arguments (`headerName`, `parameterName`, and `value`) while `Supplier` accepts no arguments.

Reply:
It would be done as constructor arguments to the `Supplier`. Something like:

```java
public static Supplier<CsrfToken> create(String headerName, String parameterName) {
    return () -> new DefaultCsrfToken(headerName, paraemterName, createTokenValue());
}
```

Reply:
How do you suggest we modify `CookieCsrfTokenRepository.loadToken()` ?
```
@Override
public CsrfToken loadToken(HttpServletRequest request) {
	Cookie cookie = WebUtils.getCookie(request, this.cookieName);
	if (cookie == null) {
		return null;
	}
	String token = cookie.getValue();
	if (!StringUtils.hasLength(token)) {
		return null;
	}
	return this.generateTokenProvider.apply(token);
}
```

Reply:
That is a good point. At this point I think it makes sense to change the API to be a `Converter<String,CsrfToken>` that allows passing in the csrf token string and the output is the `CsrfToken` implementation. When generating a new instance, the `CsrfTokenRepository` is in charge of generating the token and passing it into the `Converter<String,CsrfToken>`.

Reply:
you mean `org.springframework.core.convert.converter.Converter<S,T>` ?

So it would be like below?
```
CookieCsrfTokenRepository repo = new CookieCsrfTokenRepository();
repo.setGenerateToken((tokenStrValue) -> new DefaultCsrfToken("XSRF-TOKEN", "_csrf", tokenStrValue));
```

if so this means that `setHeaderName` and `setParameterName` will not work at all 
```
CookieCsrfTokenRepository repo = new CookieCsrfTokenRepository();
repo.setGenerateToken((tokenStrValue) -> new DefaultCsrfToken("XSRF-TOKEN", "_csrf", tokenStrValue));

repo.setParameterName("customParam"); // the token generated won't follow this value
repo.setHeaderName("customHeader"); // the token generated won't follow this value
```

and another issue is that users will be forced to maintain/hardcode default value of `parameterName` and `headerName` which were defined in both `CookieCsrfTokenRepository` and `HttpSessionCsrfTokenRepository` with default values `_csrf` and `X-CSRF-TOKEN`.

As for the current approach, it still consider `setHeaderName` and `setParameterName` from `CookieCsrfTokenRepository`
```
CookieCsrfTokenRepository repo = new CookieCsrfTokenRepository();
repo.setGenerateToken((headerName, parameterName, tokenStrValue) -> new DefaultCsrfToken(headerName, parameterName, tokenStrValue));

repo.setParameterName("customParam"); // the token generated will follow this value
repo.setHeaderName("customHeader"); // the token generated will follow this value
```

For the current approach, if users decides to maintain their own `headerName` and `parameterName` it could be done as per below
```
CookieCsrfTokenRepository repo = new CookieCsrfTokenRepository();
repo.setGenerateToken((headerName, parameterName, tokenStrValue) -> new DefaultCsrfToken("customHeader", "customParameter", tokenStrValue));

repo.setParameterName("customParam2"); // the token generated will NOT follow this value
repo.setHeaderName("customHeader2"); // the token generated will NOT follow this value
```